pipeline/nodes/basic.py

```diff
@@ -7,25 +7,149 @@
+	optional = Option(bool, default=True)
+
+	def __init__(self, *v, **kw):
+		'''
+		Sets the __name__ property to include the relevant options so that when the
+		bonobo graph is serialized as a GraphViz document, different objects can be
+		visually differentiated.
+		'''
+		super().__init__(*v, **kw)
+		self.__name__ = f'{type(self).__name__} ({self.key})'
+
+
+	@staticmethod
+	def string_to_span(value):
+		'''Parse a string value and attempt to create a corresponding `model.TimeSpan` object.'''
+		try:
+			date_from, date_to = date_cleaner(value)
+			ts = model.TimeSpan()
+			if date_from is not None:
+				ts.begin_of_the_begin = date_from.strftime("%Y-%m-%dT%H:%M:%SZ")
+			if date_to is not None:
+				ts.end_of_the_end = date_to.strftime("%Y-%m-%dT%H:%M:%SZ")
+			return ts
+		except Exception as e:
+			print('*** Unknown date format %r: %s' % (value, e))
+			return None
+
+	def __call__(self, data, *args, **kwargs):
+		if self.key in data:
+			value = data[self.key]
+			ts = self.string_to_span(value)
+			if ts is not None:
+				data['%s_span' % self.key] = ts
+				return data
+		else:
+			if not self.optional:
+				print('*** key %r is not in the data object:' % (self.key,))
+				pprint.pprint(data)
+		return NOT_MODIFIED
+
+class GroupRepeatingKeys(Configurable):
+	mapping = Option(dict)
+	drop_empty = Option(bool, default=True)
+	def __call__(self, data):
+		d = data.copy()
+		for key, mapping in self.mapping.items():
+			property_prefixes = mapping['prefixes']
+			postprocess = mapping.get('postprocess')
+			d[key] = []
+			with suppress(KeyError):
+				for i in itertools.count(1):
+					ks = ((prefix, f'{prefix}_{i}') for prefix in property_prefixes)
+					subd = {}
+					for p, k in ks:
+						subd[p] = d[k]
+						del d[k]
+					if self.drop_empty:
+						values_unset = list(map(lambda v: not bool(v), subd.values()))
+						if all(values_unset):
+							continue
+					if postprocess:
+						subd = postprocess(subd, data)
+					d[key].append(subd)
+		yield d
+
+class GroupKeys(Configurable):
+	mapping = Option(dict)
+	drop_empty = Option(bool, default=True)
+	def __call__(self, data):
+		d = data.copy()
+		for key, mapping in self.mapping.items():
+			subd = {}
+			properties = mapping['properties']
+			postprocess = mapping.get('postprocess')
+			for k in properties:
+				v = d[k]
+				del d[k]
+				if self.drop_empty and not v:
+					continue
+				subd[k] = v
+			if postprocess:
+				subd = postprocess(subd, data)
+			d[key] = subd
+		yield d
+
+class AddDataDependentArchesModel(Configurable):
+	'''
+	Set the `_ARCHES_MODEL` key in the supplied `dict` to the appropriate arches model UUID
+	and return it.
+	'''
+	models = Option()
+	def __call__(self, data, *args, **kwargs):
+		if '_LOD_OBJECT' in data:
+			obj = data['_LOD_OBJECT']
+			t = type(obj)
+			tname = t.__name__
+			if tname in self.models:
+				data['_ARCHES_MODEL'] = self.models[tname]
+				return data
+			else:
+				for c in obj._classhier:
```

Comment:
@azaroth42 Here's where I found myself wanting access to the class hierarchy. It allows the settings to specify an arches model for a set of object classes, while allowing objects to be more specific subclass instances. Not sure if that's in line with how arches works. Let's discuss.

Reply:
Arches doesn't really process the class hierarchy when it comes to models. If the data had a class of Person, and the Arches model had a class of Actor (or vice versa), it would reject it as a mismatch in the data/model ... even though it's technically correct.

That said, for other systems and other uses, access to the class hierarchy does sound like a nice feature to add.

Reply:
Ah. OK. Then I'll need to rework how to approach that. Might be a nice pattern for other uses in the future, though.